### PR TITLE
view engine fixes

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: modules/@angular/core/src/view/types.ts [hunk 1]
```diff
@@ -151,12 +151,13 @@ export enum NodeFlags {
   AfterViewChecked = 1 << 7,
   HasEmbeddedViews = 1 << 8,
   HasComponent = 1 << 9,
```
Comment (tbosch) [inline modules/@angular/core/src/view/types.ts]: Maybe remove again, and just look into `elementDef.childFlags` with `IsComponent`

File: modules/@angular/compiler/src/view_compiler_next/view_compiler.ts [hunk 2]
```diff
@@ -297,33 +297,44 @@ class ViewBuilder implements TemplateAstVisitor, LocalResolver, BuiltinConverter
...
 
     templateVisitAll(this, ast.children);
 
     const childCount = this.nodeDefs.length - nodeIndex - 1;
 
+    const compAst = ast.directives.find(dirAst => dirAst.directive.isComponent);
+    let compRendererType = o.NULL_EXPR;
+    let compView = o.NULL_EXPR;
+    if (compAst) {
+      flags |= viewEngine.NodeFlags.IsComponent;
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler_next/view_compiler.ts]: remove this again!

File: modules/@angular/core/src/view/element.ts [hunk 3]
```diff
@@ -93,29 +99,35 @@ export function elementDef(
...
-    let eventName: string;
-    if (Array.isArray(output)) {
-      [target, eventName] = output;
-    } else {
-      eventName = output;
-    }
-    outputDefs[i] = {eventName: eventName, target: target};
+    const [target, eventName] = outputs[i];
+    outputDefs[i] = {
+      type: OutputType.ElementOutput,
```
Comment (tbosch) [inline modules/@angular/core/src/view/element.ts]: Maybe introduce `OutputType.ComponentHostOutput` again here?

File: modules/@angular/core/src/view/services.ts [hunk 4]
```diff
@@ -292,21 +292,14 @@ class DebugContext_ implements DebugContext {
...
-      return asProviderData(this.elView, this.compProviderDef.index).instance;
-    }
-    return this.view.context;
+  private get elOrCompView() {
+    // Has to be done lazily as we use the DebugContext also during creation of elements...
+    return asElementData(this.elView, this.elDef.index).componentView || this.view;
   }
+  get injector(): Injector { return createInjector(this.elView, this.elDef); }
+  get component(): any { return this.elOrCompView.component; }
+  get context(): any { return this.elOrCompView.component; }
```
Comment (tbosch) [inline modules/@angular/core/src/view/services.ts]: Use `.context`

File: modules/@angular/compiler/src/view_compiler_next/view_compiler.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler_next/view_compiler.ts]: Done

File: modules/@angular/core/src/view/services.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/core/src/view/services.ts]: Done

**Review by vicb — APPROVED** [review]

File: modules/@angular/core/src/view/element.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/core/src/view/element.ts]: Decided against it.

**Pull request merged**